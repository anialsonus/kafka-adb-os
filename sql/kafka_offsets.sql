-- Test how extension synchronizes offsets with Kafka
-- This test REQUIRES a running Kafka instance. See 'kafka_test' directory for
-- instructions on how to setup one

-- start_matchignore
-- m/^NOTICE:  Kafka-ADB: Offset for partition [0-9]* is not known, and is set to default value [0-9]*$/
-- end_matchignore

-- start_ignore
DROP EXTENSION IF EXISTS kadb_fdw CASCADE;
CREATE EXTENSION kadb_fdw;

CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers 'localhost:9092',
    format 'avro'
);
-- end_ignore


-- Test: Automatic partition setup

-- start_ignore
DROP FOREIGN TABLE IF EXISTS test_kadb_fdw_table;

CREATE FOREIGN TABLE test_kadb_fdw_table(
    id INT,
    gnr TEXT,
    stu TEXT,
    ser TEXT,
    num TEXT,
    iss_by TEXT,
    iss_da INT,
    iss_plc TEXT,
    exp_dat INT,
    det_dat TEXT,
    r_obj INT,
    crt_on INT,
    upd_on INT,
    dsc TEXT NULL,
    iss_id TEXT,
    vrf_stu TEXT,
    vrf_on INT,
    sys_op INT
)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'kadb_fdw_test',
    k_consumer_group 'test_consumer_group',
    k_seg_batch '5',
    k_timeout_ms '2000'
);
-- end_ignore

SELECT id, gnr, dsc FROM test_kadb_fdw_table ORDER BY id, gnr, dsc;


-- Test: kadb.commit_offsets()

-- start_ignore
DROP FOREIGN TABLE IF EXISTS test_kadb_fdw_table;

CREATE FOREIGN TABLE test_kadb_fdw_table(
    id INT,
    gnr TEXT,
    stu TEXT,
    ser TEXT,
    num TEXT,
    iss_by TEXT,
    iss_da INT,
    iss_plc TEXT,
    exp_dat INT,
    det_dat TEXT,
    r_obj INT,
    crt_on INT,
    upd_on INT,
    dsc TEXT NULL,
    iss_id TEXT,
    vrf_stu TEXT,
    vrf_on INT,
    sys_op INT
)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'kadb_fdw_test',
    k_consumer_group 'test_consumer_group',
    k_seg_batch '5',
    k_timeout_ms '2000'
);

SELECT id, gnr, dsc FROM test_kadb_fdw_table ORDER BY id, gnr, dsc;
-- end_ignore

SELECT prt, off FROM kadb.offsets WHERE ftoid = 'test_kadb_fdw_table'::regclass::oid ORDER BY prt;

SELECT kadb.commit_offsets('test_kadb_fdw_table'::regclass::oid);

-- kafka-consumer-groups succeeds only when offsets are committed.
-- Otherwise, it fails: "Consumer group 'test_consumer_group' does not exist."
\! docker exec broker kafka-consumer-groups --bootstrap-server localhost:9092 --describe --group test_consumer_group 2>/dev/null | sort
