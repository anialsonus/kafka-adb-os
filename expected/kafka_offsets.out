-- Test how extension synchronizes offsets with Kafka
-- This test REQUIRES a running Kafka instance. See 'kafka_test' directory for
-- instructions on how to setup one
-- start_matchignore
-- m/^NOTICE:  Kafka-ADB: Offset for partition [0-9]* is not known, and is set to default value [0-9]*$/
-- end_matchignore
-- start_ignore
DROP EXTENSION IF EXISTS kadb_fdw CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to server test_kadb_fdw_server
drop cascades to foreign table test_kadb_fdw_table
CREATE EXTENSION kadb_fdw;
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers 'localhost:9092',
    format 'avro'
);
-- end_ignore
-- Test: Automatic partition setup
-- start_ignore
DROP FOREIGN TABLE IF EXISTS test_kadb_fdw_table;
NOTICE:  foreign table "test_kadb_fdw_table" does not exist, skipping
CREATE FOREIGN TABLE test_kadb_fdw_table(
    id INT,
    gnr TEXT,
    stu TEXT,
    ser TEXT,
    num TEXT,
    iss_by TEXT,
    iss_da INT,
    iss_plc TEXT,
    exp_dat INT,
    det_dat TEXT,
    r_obj INT,
    crt_on INT,
    upd_on INT,
    dsc TEXT NULL,
    iss_id TEXT,
    vrf_stu TEXT,
    vrf_on INT,
    sys_op INT
)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'kadb_fdw_test',
    k_consumer_group 'test_consumer_group',
    k_seg_batch '5',
    k_timeout_ms '2000'
);
-- end_ignore
SELECT id, gnr, dsc FROM test_kadb_fdw_table ORDER BY id, gnr, dsc;
NOTICE:  Kafka-ADB: Offset for partition 0 is not known, and is set to default value 0
NOTICE:  Kafka-ADB: Offset for partition 1 is not known, and is set to default value 0
NOTICE:  Kafka-ADB: Offset for partition 2 is not known, and is set to default value 0
NOTICE:  Kafka-ADB: Offset for partition 3 is not known, and is set to default value 0
   id   |          gnr          | dsc 
--------+-----------------------+-----
 251306 | OLD_BIRTH_CERTIFICATE | 
 251306 | OLD_BIRTH_CERTIFICATE | 
 251306 | OLD_BIRTH_CERTIFICATE | 
 251306 | OLD_BIRTH_CERTIFICATE | 
 251306 | OLD_BIRTH_CERTIFICATE | 
 251306 | OLD_BIRTH_CERTIFICATE | 
 251310 | RF_FOREIGNER_ID_DOC   | 
 251310 | RF_FOREIGNER_ID_DOC   | 
 251310 | RF_FOREIGNER_ID_DOC   | 
 251310 | RF_FOREIGNER_ID_DOC   | 
 251310 | RF_FOREIGNER_ID_DOC   | 
 251310 | RF_FOREIGNER_ID_DOC   | 
 251313 | SEAMAN_BOOK           | 
 251313 | SEAMAN_BOOK           | 
 251313 | SEAMAN_BOOK           | 
 251313 | SEAMAN_BOOK           | 
 251313 | SEAMAN_BOOK           | 
 251313 | SEAMAN_BOOK           | 
(18 rows)

-- Test: kadb.commit_offsets()
-- start_ignore
DROP FOREIGN TABLE IF EXISTS test_kadb_fdw_table;
CREATE FOREIGN TABLE test_kadb_fdw_table(
    id INT,
    gnr TEXT,
    stu TEXT,
    ser TEXT,
    num TEXT,
    iss_by TEXT,
    iss_da INT,
    iss_plc TEXT,
    exp_dat INT,
    det_dat TEXT,
    r_obj INT,
    crt_on INT,
    upd_on INT,
    dsc TEXT NULL,
    iss_id TEXT,
    vrf_stu TEXT,
    vrf_on INT,
    sys_op INT
)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'kadb_fdw_test',
    k_consumer_group 'test_consumer_group',
    k_seg_batch '5',
    k_timeout_ms '2000'
);
SELECT id, gnr, dsc FROM test_kadb_fdw_table ORDER BY id, gnr, dsc;
NOTICE:  Kafka-ADB: Offset for partition 0 is not known, and is set to default value 0
NOTICE:  Kafka-ADB: Offset for partition 1 is not known, and is set to default value 0
NOTICE:  Kafka-ADB: Offset for partition 2 is not known, and is set to default value 0
NOTICE:  Kafka-ADB: Offset for partition 3 is not known, and is set to default value 0
   id   |          gnr          | dsc 
--------+-----------------------+-----
 251306 | OLD_BIRTH_CERTIFICATE | 
 251306 | OLD_BIRTH_CERTIFICATE | 
 251306 | OLD_BIRTH_CERTIFICATE | 
 251306 | OLD_BIRTH_CERTIFICATE | 
 251306 | OLD_BIRTH_CERTIFICATE | 
 251306 | OLD_BIRTH_CERTIFICATE | 
 251310 | RF_FOREIGNER_ID_DOC   | 
 251310 | RF_FOREIGNER_ID_DOC   | 
 251310 | RF_FOREIGNER_ID_DOC   | 
 251310 | RF_FOREIGNER_ID_DOC   | 
 251310 | RF_FOREIGNER_ID_DOC   | 
 251310 | RF_FOREIGNER_ID_DOC   | 
 251313 | SEAMAN_BOOK           | 
 251313 | SEAMAN_BOOK           | 
 251313 | SEAMAN_BOOK           | 
 251313 | SEAMAN_BOOK           | 
 251313 | SEAMAN_BOOK           | 
 251313 | SEAMAN_BOOK           | 
(18 rows)

-- end_ignore
SELECT prt, off FROM kadb.offsets WHERE ftoid = 'test_kadb_fdw_table'::regclass::oid ORDER BY prt;
 prt | off 
-----+-----
   0 |   2
   1 |   1
   2 |   0
   3 |   3
(4 rows)

SELECT kadb.commit_offsets('test_kadb_fdw_table'::regclass::oid);
 commit_offsets 
----------------
 
(1 row)

-- kafka-consumer-groups succeeds only when offsets are committed.
-- Otherwise, it fails: "Consumer group 'test_consumer_group' does not exist."
\! docker exec broker kafka-consumer-groups --bootstrap-server localhost:9092 --describe --group test_consumer_group 2>/dev/null | sort

GROUP               TOPIC           PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             CONSUMER-ID     HOST            CLIENT-ID
test_consumer_group kadb_fdw_test   0          2               2               0               -               -               -
test_consumer_group kadb_fdw_test   1          1               1               0               -               -               -
test_consumer_group kadb_fdw_test   2          0               0               0               -               -               -
test_consumer_group kadb_fdw_test   3          3               3               0               -               -               -
